core/services/keeper/registry_synchronizer_process_logs.go

```diff
@@ -0,0 +1,152 @@
+package keeper
+
+import (
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/log"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+func (rs *RegistrySynchronizer) processLogs(head models.Head) {
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+	go rs.handleSyncRegistryLogs(head, wg.Done)
+	go rs.handleUpkeepCanceledLogs(head, wg.Done)
+	go rs.handleUpkeepRegisteredLogs(head, wg.Done)
+	wg.Wait()
+}
+
+func (rs *RegistrySynchronizer) handleSyncRegistryLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbSyncRegistry.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
```

Comment:
isn't there a risk of skipping logs in a situation of a back-and-forth reorg?
- we receive head A (where parent is X) and logs for it, but they are not yet confirmed
- a head B is received (where parent is also X) and logs for A are now confirmed
- but logs for A are now for invalid chain, so are skipped
- head A' is received where parent is A, but we have already skipped the logs for it - instead ideally they should be sent now

Reply:
@PiotrTrzpil If the value of `minConfirmations` is low, say 2, then yes this situation absolutely exists. If the valuable is something more reasonable for predicting finality then we shouldn't encounter this problem. The algo basically says this...

* minConfirmations is 6
* The logs in question are in block A, height 10
* The current head is block G, 16
* If block A is part of the chain beginning with block G, then those logs are confirmed and can be broadcast
* If block A is _not_ part of the chain beginning with G, then those logs are sufficiently old we can assume they will never be confirmed and we can discard them

Reply:
ok - makes sense 👍🏼

Reply:
Wasn't @spooktheducks working on building this into the log broadcaster?

Reply:
@sam, @PiotrTrzpil is now! But it won't be ready until post-audit